tinygrad/runtime/ops_hip.py

```diff
@@ -16,53 +22,90 @@
 class RawHIPBuffer(RawBufferCopyInOut):
   def __init__(self, size, dtype): super().__init__(size, dtype, hip.hipMalloc(size * dtype.itemsize))
   def __del__(self): hip.hipFree(self._buf)
-  def _copyin(self, x:np.ndarray): hip.hipMemcpyAsync_htod(self._buf, x.ctypes.data, self.size * self.dtype.itemsize, 0)
-  def _copyout(self, x:np.ndarray): hip.hipMemcpy_dtoh(x.ctypes.data, self._buf, self.size * self.dtype.itemsize)
+  def _copyin(self, x:np.ndarray): hip.hipMemcpy(self._buf, x.ctypes.data_as(ctypes.c_void_p), self.size * self.dtype.itemsize, hip.hipMemcpyHostToDevice)
+  def _copyout(self, x:np.ndarray): hip.hipMemcpy(x.ctypes.data_as(ctypes.c_void_p), self._buf, self.size * self.dtype.itemsize, hip.hipMemcpyDeviceToHost)
 
 class HIPProgram:
   def __init__(self, name:str, prg:str, binary=False):
     try:
       if not binary:
-        prog = hip.hiprtcCreateProgram(prg, name, [], [])
-        device_properties = hip.hipGetDeviceProperties(hip.hipGetDevice())
-        hip.hiprtcCompileProgram(prog, [f'--offload-arch={device_properties.gcnArchName}'])
-        prg = hip.hiprtcGetCode(prog)
+        if HIP_CPU:
+          prg = "#include <hip/hip_runtime.h>\n" +  prg
+        else:
+          prog = hip.hiprtcCreateProgram(prg, name, [], [])
+          device_properties = hip.hipGetDeviceProperties(hip.hipGetDevice())
+          hip.hiprtcCompileProgram(prog, [f'--offload-arch={device_properties.gcnArchName}'])
+          prg = hip.hiprtcGetCode(prog)
     except Exception as e:
       if DEBUG >= 3: print("FAILED TO BUILD", prg)
       raise e
     if DEBUG >= 5:
       asm = early_exec((["/opt/rocm/llvm/bin/llvm-objdump", '-d', '-'], prg))
       print('\n'.join([x for x in asm.decode('utf-8').split("\n") if 's_code_end' not in x]))
-
-    module = hip.hipModuleLoadData(prg)
-    self.prg = hip.hipModuleGetFunction(module, name)
+    if HIP_CPU:
+      import tempfile, hashlib, os, subprocess
+      fn = f"{tempfile.gettempdir()}/clang_{hashlib.md5(prg.encode('utf-8')).hexdigest()}.so"
+      if not os.path.exists(fn):
+        subprocess.check_output(args=('clang -g -std=c++20 --stdlib=libstdc++ -shared -O2 -Wall -Werror -x c++ -ltbb -ltbbmalloc -fPIC --rtlib=compiler-rt - -o '+fn+'.tmp').split(), input=prg.encode('utf-8'))
```

Comment:
this is a race condition - add process id to tmp file name - see `ops_clang.py` to see how it is done:

https://github.com/tinygrad/tinygrad/blob/a293c18d345a45ff7a315c8fb5da17437c129949/tinygrad/runtime/ops_clang.py#L38